[13.0][MIG] sale_invoice_plan

kittiu opened the pull request:
Standard migration

- [x] Merge changes in #967 to this PR
- [x] included this fixed on v12 https://github.com/OCA/sale-workflow/pull/1130

kittiu committed:
[ADD][12.0] sale_invoice_plan

oca-travis committed:
[UPD] Update sale_invoice_plan.pot

OCA-git-bot committed:
[UPD] README.rst

liweijie0812 committed:
Added translation using Weblate (Chinese (Simplified))

liweijie0812 committed:
Translated using Weblate (Chinese (Simplified))

Currently translated at 100.0% (56 of 56 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_invoice_plan
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_invoice_plan/zh_CN/

OCA-git-bot committed:
[UPD] README.rst

sudhir-erpharbor commented on sale_invoice_plan/models/sale.py:
```diff
@@ -0,0 +1,215 @@
+from odoo import models, fields, api, _
+from odoo.exceptions import UserError, ValidationError
+from odoo.addons import decimal_precision as dp
+from odoo.tools.float_utils import float_compare, float_round as round
```
Rename round variable to something else as this is a built-in method of python.

Travis: sale_invoice_plan/models/sale.py:7: [W0622(redefined-builtin), ] Redefining built-in 'round'

sudhir-erpharbor commented on sale_invoice_plan/wizard/sale_make_invoice_advance.py:
```diff
@@ -0,0 +1,15 @@
+        invoice_plan_id = self._context.get('invoice_plan_id')
+        if invoice_plan_id:
+            plan = self.env['sale.invoice.plan'].browse(invoice_plan_id)
+            plan.invoice_ids += invoice
```
It should be "invoice_move_ids" instead of "invoice_ids"

Test case error:
AttributeError: 'sale.invoice.plan' object has no attribute 'invoice_ids'

sudhir-erpharbor commented on sale_invoice_plan/tests/test_sale_invoice_plan.py:
```diff
@@ -0,0 +1,184 @@
+        # Create one invoice
+        make_wizard = self.env['sale.make.planned.invoice'].create({})
+        make_wizard.with_context(ctx).create_invoices_by_plan()
+        invoices = self.so_service.invoice_ids
```
invoice_move_ids

sudhir-erpharbor commented on sale_invoice_plan/tests/test_sale_invoice_plan.py:
```diff
@@ -0,0 +1,184 @@
+        ctx['all_remain_invoices'] = True
+        make_wizard.with_context(ctx).create_invoices_by_plan()
+        # Valid number of invoices
+        invoices = self.so_service.invoice_ids
```
invoice_move_ids

kittiu replied:
invoice_ids is already correct.

sudhir-erpharbor commented on sale_invoice_plan/tests/test_sale_invoice_plan.py:
```diff
@@ -0,0 +1,184 @@
+        wizard = self.env['sale.make.planned.invoice'].create({})
+        wizard.with_context(ctx).create_invoices_by_plan()
+        # Valid number of invoices, including advance
+        invoices = self.so_service.invoice_ids
```
invoice_move_ids

sudhir-erpharbor commented on sale_invoice_plan/tests/test_sale_invoice_plan.py:
```diff
@@ -0,0 +1,184 @@
+        wizard = self.env['sale.make.planned.invoice'].create({})
+        wizard.with_context(ctx).create_invoices_by_plan()
+        # Valid total quantity of invoice = 10 units
+        invoices = self.so_product.invoice_ids
```
invoice_move_ids

kittiu replied:
invoice_ids is available for sale.order

sudhir-erpharbor commented on sale_invoice_plan/models/sale.py:
```diff
@@ -0,0 +1,215 @@
+        invoice_plan_id = self._context.get('invoice_plan_id')
+        if invoice_plan_id:
+            plan = self.env['sale.invoice.plan'].browse(invoice_plan_id)
+            invoices = self.env['account.invoice'].browse(inv_ids)
```
I think it should be account.move since invoice object is removed.

sudhir-erpharbor commented on sale_invoice_plan/README.rst:
```diff
@@ -0,0 +1,113 @@
+
+|maintainer-kittiu| 
+
+This module is part of the `OCA/sale-workflow <https://github.com/OCA/sale-workflow/tree/12.0/sale_invoice_plan>`_ project on GitHub.
```
13.0 in the url

sudhir-erpharbor commented on sale_invoice_plan/README.rst:
```diff
@@ -0,0 +1,113 @@
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/sale-workflow/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed
+`feedback <https://github.com/OCA/sale-workflow/issues/new?body=module:%20sale_invoice_plan%0Aversion:%2012.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
13.0

kittiu replied:
I removed README.rst, it will be created auto by ocabot when merged.

sudhir-erpharbor reviewed (commented):
Important changes

Commit:
Added translation using Weblate (Dutch)

Commit:
Translated using Weblate (Dutch)

Currently translated at 100.0% (56 of 56 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_invoice_plan
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_invoice_plan/nl_NL/

kittiu commented:
Ready fo review.

sudhir-erpharbor commented:
Please fix travis.

sale_invoice_plan/__init__.py:1: [C7902(missing-readme), ] Missing ./README.rst file. Template here: https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

sudhir-erpharbor commented on sale_invoice_plan/models/sale.py:
```diff
@@ -0,0 +1,228 @@
+                if float_compare(plan_qty, line.quantity, prec) == 1:
+                    raise ValidationError(
+                        _('Plan quantity: %s, exceed invoiceable quantity: %s'
+                          '\nProduct should be deliver  ed before invoice') %
```
delivered. There is a space between "deliver" & "ed"

kittiu replied:
Thank you, fixed.

kittiu commented:
The error in on this line,
`sale_invoice_plan/wizard/sale_make_planned_invoice.py:3:1: F401 'odoo.api' imported but unused
`
Fixed.

kittiu commented:
@sudhir-erpharbor thank you for your reviewed, all fixed.
Thanks!

kittiu commented:
The remaining problem is on runbot, which I think it might be runbot config and not related to this module.
```
2019-10-08 06:27:53,633 150 [1;33m[1;49mWARNING[0m openerp_template odoo.addons.phone_validation.tools.phone_validation: The `phonenumbers` Python module is not installed, contact numbers will not be verified. Please install the `phonenumbers` Python module. 
```

sudhir-erpharbor commented:
Code review LGTM

kittiu commented:
> Code review LGTM

:+1: please approve if you see appropriate. Thank you for your reviews.

kittiu commented:
@Saran440 please can you help review this.

bealdav commented on sale_invoice_plan/models/sale.py:
```diff
@@ -0,0 +1,228 @@
+from odoo.tools.float_utils import float_compare, float_round
+
+
+class SaleOder(models.Model):
```
s/SaleOder/SaleOrder

kittiu replied:
@bealdav thanks for reviewing. I am not sure what your comment is about :p

bealdav replied:
oh, it's just a typo error.
You wrote `SaleOder`, I suppose you would write `SaleOrder` :)

kittiu committed:
[12.0][IMP] sale_invoice_plan improvements

1. Plan date now passed to invoice date
2. Allow edit invoice plan, if no invoice created yet (invoice_count = 0)
3. Add new view Sales Invoice Plan

kittiu commented:
I had misc enhancement in [12.0] https://github.com/OCA/sale-workflow/pull/967
I will wait until that one is merged, and will add the same change here.

Saran440 commented:
@kittiu Functional Test : Found 2 bugs
**1.Invoice Plan calculate only first, when post invoice**
step :
- Create sales and invoice plan 4 installment
- Confirm and receive order 1/4
![Selection_003](https://user-images.githubusercontent.com/20896369/67467757-8c7a2d80-f673-11e9-949d-65ef74cba7b5.png)
- Create Invoice by Plan -> Create Next Invoice
![Selection_004](https://user-images.githubusercontent.com/20896369/67468227-525d5b80-f674-11e9-8bf2-59c79e8174e1.png)
- Post invoice
![Selection_006](https://user-images.githubusercontent.com/20896369/67468863-78cfc680-f675-11e9-97ac-7efc93547fc7.png)
-  Sale invoice plan will back to first
![Selection_005](https://user-images.githubusercontent.com/20896369/67468806-59389e00-f675-11e9-8472-5873238e3259.png)

**2.Due Date and Invoice Date in Invoices mistake**
step :
- Create normal process and use invoice plan
- Click create all remaining invoices
![Selection_007](https://user-images.githubusercontent.com/20896369/67469700-c436a480-f676-11e9-8773-a84deab9927e.png)
- Invoice date should be depend plan date and due date shouldn't same as.
![Selection_008](https://user-images.githubusercontent.com/20896369/67469658-b4b75b80-f676-11e9-8d4b-65e25e5d9ad5.png)

Martronic-SA committed:
Added translation using Weblate (French)

Martronic-SA committed:
Translated using Weblate (French)

Currently translated at 92.9% (52 of 56 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_invoice_plan
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_invoice_plan/fr/

oca-travis committed:
[UPD] Update sale_invoice_plan.pot

OCA-git-bot committed:
sale_invoice_plan 12.0.2.0.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_invoice_plan
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_invoice_plan/

newtratip commented:
Red CI

AaronHForgeFlow commented:
If I select and advance invoice int he plan:
![image](https://user-images.githubusercontent.com/19620251/80579382-37481b80-8a0a-11ea-98a1-69fcc4cab997.png)

I cannot see this button until i send some quantity, I think this is not correct:
![image](https://user-images.githubusercontent.com/19620251/80579357-2c8d8680-8a0a-11ea-862e-39cc7395d5fb.png)

AaronHForgeFlow commented:
Also, if I select advance payment in the invoice plan the first invoice I generate using plan invoice will be generate correctly, but not the second one. See the example below:

I select advance invoice for 50% of the total and then the next invoice is for another 40%:
![image](https://user-images.githubusercontent.com/19620251/80580538-fa7d2400-8a0b-11ea-8565-8dee8cb6e073.png)

The first invoice is created correctly, however the second one is not, it generates again another down payment for the 50%, which is not correct:

![image](https://user-images.githubusercontent.com/19620251/80580445-d3beed80-8a0b-11ea-9bb8-7749c68cadb4.png)

kittiu commented:
@AaronHForgeFlow thank you for your review. I will check that out.

AaronHForgeFlow commented:
Thank you @kittiu  Perhaps it is that I am not understanding correctly the concept of advance payment. Imagine I have this situation where I invoice in advance 50% of the total and then 40% the next month:
![image](https://user-images.githubusercontent.com/19620251/80583869-1df69d80-8a11-11ea-827e-a98002104f03.png)

The advance invoice is created correctly for half price of the sales order. However, when creating the next invoice it creates an invoice line for the 40% (correct) but it is including a credit note discount for the 40% of the advance (see image below):
![image](https://user-images.githubusercontent.com/19620251/80584009-5b5b2b00-8a11-11ea-853f-3e46cc8803f8.png)

I was expecting just the first two invoice lines to be created, but not that extra line. Maybe am I missing something here, if you could shed some light here it would be awesome. Thanks for the work.

AaronHForgeFlow reviewed (commented):
I tested now the module without choosing the "Advance On 1st" Option. I selected 3 Installments. I created the first invoice based on plan and I posted it. However, the invoice plan was not updated:
![image](https://user-images.githubusercontent.com/19620251/80588146-14246880-8a18-11ea-9fa9-333945c3a676.png)

Additionally, if I create another invoice from the plan, then it works, like it were the first one, and I get something like this:
![image](https://user-images.githubusercontent.com/19620251/80588485-bd6b5e80-8a18-11ea-8536-6d58e1f57d77.png)

Sorry for all the separate reviews :)

kittiu commented:
@AaronHForgeFlow thank you so much for the detailed test. I have made quite a number of fixes. Can you test again please?

The concept is this,
![image](https://user-images.githubusercontent.com/1973598/80796484-77093180-8bc9-11ea-83b4-82f48dff2f14.png)

Also, as you say here,
> The advance invoice is created correctly for half price of the sales order. However, when creating the next invoice it creates an invoice line for the 40% (correct) but it is including a credit note discount for the 40% of the advance (see image below):

I think the standard Odoo advance also works this way, having the minus value of its part. And I just try to reuse it.

kittiu commented:
> If I select and advance invoice int he plan:
> ![image](https://user-images.githubusercontent.com/19620251/80579382-37481b80-8a0a-11ea-98a1-69fcc4cab997.png)
> 
> I cannot see this button until i send some quantity, I think this is not correct:
> ![image](https://user-images.githubusercontent.com/19620251/80579357-2c8d8680-8a0a-11ea-862e-39cc7395d5fb.png)

Just this one, I guess this should be bug since v12 also, did you have chance to test it?

AaronHForgeFlow commented:
> 
> Just this one, I guess this should be bug since v12 also, did you have chance to test it?

Yes, and in v12 it was happening the same thing. Thanks for the quick update. :)

AaronHForgeFlow commented:
> I think the standard Odoo advance also works this way, having the minus value of its part. And I just try to reuse it.

Yes. Please disregard that comment.

AaronHForgeFlow reviewed (approved):
Tested on local. All issues are solved now. :+1:

kittiu commented:
@Saran440 your issue is also fixed. Would you approve?

kittiu commented:
Not really sure why runbot is failed, there is no error.

AaronHForgeFlow commented:
@kittiu 

I think it is this warning:
`2020-05-01 16:10:12,740 165 WARNING openerp_test odoo.addons.base.models.decimal_precision: Deprecated call to decimal_precision.get_precision(<application>), use digits=<application> instead`

kittiu commented:
@AaronHForgeFlow :+1:

kittiu commented:
@AaronHForgeFlow can merge?

kittiu commented:
@AaronHForgeFlow @HviorForgeFlow 
Trying to clear all the sale/purchase invoice plan up to v13.

AaronHForgeFlow commented:
For me it is ok to move forward.  cc @JordiBForgeFlow

kittiu commented:
@OCA/crm-sales-marketing-maintainers can we merge this?

Saran440 reviewed (approved):
Code Reviewed.

you should seperate 2 commits
- [IMP] $MODULE: black, isort, prettier for pre-commit pattern
- [MIG] xxx for easy reviewed

kittiu committed:
[IMP] : black, isort, prettier

kittiu committed:
[13.0][MIG] sale_invoice_plan

kittiu commented:
> Code Reviewed.
> 
> you should seperate 2 commits
> 
> * [IMP] $MODULE: black, isort, prettier for pre-commit pattern
> * [MIG] xxx for easy reviewed

Done

HviorForgeFlow reviewed (changes requested):
Functional Review:

Using first month in Advance when confirming Quotation:
![image](https://user-images.githubusercontent.com/32061121/92098002-01aca800-edd9-11ea-82e2-0d548a3fd6f3.png)

Trying to fix that I decided to remove advance installment to move forward. I notice when you remove an installment Create Invoice Plan Menu button disappears. I've tried to solve it removing Invoice Plan and recreating it once Sale Order is confirm, same behavior I can't create the invoice plan. So regarding there is no specified group security, everyone can modify the invoice plan model, I've got stuck went trying to fix. Same happens if I edit Invoice Plan before Confirm the Order.

Does Invoice Plan should be immutable once created? For example, I want to create a non linear invoice plan reusing a wizard creation's one and modify percentages.

@kittiu

kittiu commented:
@HviorForgeFlow thank you for your review.

I can't produce the behavior you mentioned. What kind of product you use, is it product or service? If you use product, only if product delivered, then the "Create Invoice Plan" button will be shown.

I try with both product (make sure it delivered) and service. And you can remove or recreate invoice plan once it is confirmed. This is intentional. Only when some invoice(s) is created, it will be immutable.

HviorForgeFlow commented:
Ok. When Invoice Policy is invoice what is delivered it seems to work fine, not mention in readme, I had the problems with the other policy. I'm still having problem when use Advance on 1st invoice. Error in previous image I uploaded.

kittiu commented:
Error because you didnt input the percentage for advance yet.

kittiu commented:
I am not sure I understand your problem yet. ;) may ne you can elaborate a bit.

HviorForgeFlow commented:
![sale_invoice_plan_error](https://user-images.githubusercontent.com/32061121/92248122-7ca0bc00-eec8-11ea-8fdb-d601c010c46b.gif)

kittiu commented:
That is because the advance is 0 percent. You should edit it first. For example make it 10 percent.

rousseldenis commented:
@kittiu Is this ready ?

kittiu commented:
Yes, very much. :)

kittiu commented:
@rousseldenis, can merged? :)

kittiu commented:
> @rousseldenis, can merged? :)
@rousseldenis, would you?

rousseldenis commented:
/ocabot merge patch

rousseldenis commented on sale_invoice_plan/views/sale_view.xml:
```diff
@@ -0,0 +1,183 @@
+                    />
+                    <field
+                        name="invoice_plan_ids"
+                        context="{'tree_view_ref': 'view_sale_invoice_plan_tree'}"
```
@kittiu Could you make a fix for this. This generates a warning on view load. 

See https://travis-ci.com/github/OCA/sale-workflow/jobs/494482747#L1767

Thanks !
